mxfusion/inference/score_function.py

```diff
@@ -0,0 +1,187 @@
+        :type parameters: {Variable: mxnet.ndarray.ndarray.NDArray or
+            mxnet.symbol.symbol.Symbol}
+        :returns: the outcome of the inference algorithm
+        :rtype: mxnet.ndarray.ndarray.NDArray or mxnet.symbol.symbol.Symbol
+        """
+        knowns = data.copy()
+        knowns.update(parameters)
+        knowns.update(constants)
+
+        # Sample q(z | lambda)
+        samples = self.posterior.draw_samples(
+            F=F, conditionals=knowns, num_samples=self.num_samples,
+            constants=constants)
+        knowns.update(samples)
+
+        q_z_lambda = self.posterior.compute_log_prob(
+            F=F, targets=knowns, constants=constants)
+
+        p_x_z = self.model.compute_log_prob(
+            F=F, targets=knowns, constants=constants)
+
+        gradient_theta = F.mean(p_x_z - F.stop_gradient(q_z_lambda), axis=0)
+        posterior_rvs = [v for v in self.posterior.variables.values() if v.type is VariableType.RANDVAR]
+        f_list = []
+        grad_list = []
+        for i, v in enumerate(posterior_rvs):
+            model_v = self.model[v]
+
+            q_i_varset = self._extract_descendant_blanket_params(self.posterior, v)
+            q_i_params = {key:val for key,val in knowns.items() if key in q_i_varset}
+            q_i = self.posterior.compute_log_prob(F=F,
+                  targets=q_i_params,
+                  constants=knowns)
+
+            p_i_varset = self._extract_descendant_blanket_params(self.model, model_v)
+            p_i_params = {key:val for key,val in knowns.items() if key in p_i_varset}
+            p_i = self.model.compute_log_prob(F=F,
+                  targets=p_i_params,
+                  constants=constants)
+
+            # TODO Remove this hack one day when MXNet doesn't have a bug?
+            # Need to stop the gradient of p_i manually, for some reason it doesn't like
+            # being used directly in this computation. Possibly only when p_i == p_x_z but
+            # that is unconfirmed.
+            f_i = q_i * (p_i.asscalar() - q_i.asscalar())
+
+            # f_i = q_i * F.stop_gradient(p_i - q_i)
+            f_list.append(F.expand_dims(f_i, axis=0))
+
+            # With control variate?
+            # h[i] = q_i
+            # a = F.sum(f[i]) # covariance equation
+
+            grad_i = F.mean(f_i, axis=0)
+            grad_list.append(F.expand_dims(grad_i, axis=0))
+
+        f = F.concat(*f_list, dim=0)
+        grad = F.concat(*grad_list, dim=0)
+
+        gradient_lambda = F.sum(grad)
+
+
+        # Robbins-Monro sequence??
+        gradient_log_L = gradient_lambda + gradient_theta
+
+        return -gradient_theta, -gradient_log_L
+
+    def _extract_descendant_blanket_params(self, graph, node):
+        """
+        Returns a set of the markov blankets of all of the descendents of the node in the graph, mapped to their parameter form.
+        """
+        if node.graph != graph.components_graph:
```

Comment:
Is this the expected way in Python to check two reference points to the same data structure?

Reply:
As long as the __eq__ operator isn't overridden by the class then yeah, it is the same as call id(object) == id(object2). So this is fine I think. Could do node.graph is graph.components_graph if we wanted to always use id() comparison.